File: src/no/kristiania/pgr301/minesweeper/Minefield.java
package no.kristiania.pgr301.minesweeper;

public class Minefield {
    private final String[] input;

    public Minefield(String[] input) {
        this.input = input;
    }

    public String[] getHints() {
        String[] result = new String[input.length];
        for (int row = 0, inputLength = input.length; row < inputLength; row++) {
            StringBuilder line = new StringBuilder();
            for (int col = 0; col < input[row].length(); col++) {
                if (hasMine(row, col)) {
                    line.append("*");
                } else {
                    int neighbourMines = 0;
                    if (hasMine(row, col-1)) {
                        neighbourMines = 1;
                    } else if (hasMine(row, col+1)) {
                        neighbourMines = 1;
                    } else if (hasMine(row - 1, col)) {
                        neighbourMines = 1;
                    } else if (hasMine(row + 1, col)) {
                        neighbourMines = 1;
                    }
                    line.append(neighbourMines);
                }
            }
            result[row] = line.toString();
        }
        return result;
    }

    private boolean hasMine(int row, int col) {
        if (row < 0 || input.length <= row) {
            return false;
        }
        if (col < 0 || input[row].length() <= col) {
            return false;
        }
        return input[row].charAt(col) == '*';
    }
}

File: Minesweeper/Minefield.cs
namespace Minefield;

public class Minefield
{
    private string[] input;

    public Minefield(string[] input)
    {
        this.input = input;
    }

    public string[] CalculateHints()
    {
        var result = new string[input.Length];
        for (int row = 0; row < input.Length; row++)
        {
            result[row] = "";
            for (int col = 0; col < input[row].Length; col++)
            {
                if (CellHasMine(row, col))
                {
                    result[row] += "*";
                }
                else
                {
                    result[row] += CountMines(row, col);
                }
            }
        }
        return result;
    }

    private int CountMines(int row, int col)
    {
        var mines = 0;

        if (CellHasMine(row - 1, col - 1)) mines += 1;
        if (CellHasMine(row - 1, col)) mines += 1;
        if (CellHasMine(row - 1, col + 1)) mines += 1;
        if (CellHasMine(row, col + 1)) mines += 1;
        if (CellHasMine(row, col - 1)) mines += 1;
        if (CellHasMine(row + 1, col - 1)) mines += 1;
        if (CellHasMine(row + 1, col)) mines += 1;
        if (CellHasMine(row + 1, col + 1)) mines += 1;
        return mines;
    }

    private bool CellHasMine(int row, int col)
    {
        if (row < 0 || input.Length <= row) return false;
        if (col < 0 || input[row].Length <= col) return false;
        return input[row][col] == '*';
    }
}